src/mlpack/methods/ann/augmented/ham_unit_impl.hpp

```diff
@@ -0,0 +1,96 @@
+/**
+ * @file ham_unit_impl.hpp
+ * @author Konstantin Sidorov
+ *
+ * Implementation of HAMUnit class
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_AUGMENTED_HAM_UNIT_IMPL_HPP
+#define MLPACK_METHODS_ANN_AUGMENTED_HAM_UNIT_IMPL_HPP
+
+#include "ham_unit.hpp"
+
+namespace mlpack {
+namespace ann /* Artificial Neural Network */ {
+namespace augmented /* Augmented neural network */ {
+
+template<typename E, typename J, typename S, typename W, typename C>
+HAMUnit<E, J, S, W, C>::HAMUnit(size_t memorySize,
+                 size_t memoryDim,
+                 E& embed,
+                 J& join,
+                 S& search,
+                 W& write,
+                 C& controller)
+  : memorySize(memorySize), memoryDim(memoryDim),
+    search(search), embed(embed), controller(controller),
+    memory(TreeMemory<double, J, S>(memorySize, memoryDim, join, write))
+{
+  this->t = 0;
+}
+
+template<typename E, typename J, typename S, typename W, typename C>
+arma::vec HAMUnit<E, J, S, W, C>::Attention()
```

Comment:
To you mind to pas the output by reference here, to be consistent with the rest of the codebase?